Spans's parent must be passed as Context instead of Span(Context)s.

Oberon00 opened the pull request:
* Fixes #510 "Consider using Context as the unique way to specify parenthood".
* I think this also goes a long way for fixing #867 "How does Span gets entries from CorrelationContext/Baggage".
* Helps a bit for telemetry/oteps#58 "Unseal SpanContext".

## Changes

~~Before this PR, a span's parent was logically a `SpanContext`. After this PR, it is a full `Context`:~~
* A full `Context` is the only way to specify a parent of a span. `SpanContext` or even `Span` are not possible anymore.
* ~~Readable spans MUST provide a way to retrieve the full parent `Context`.~~

</del>

## Prototype

There is an implementation of this in Java: open-telemetry/opentelemetry-java#1611 Note: The prototype does not implement the Context additional parameter to OnStart, but that would be trivial to add in Java.

## Rationale

This is required to actually transport arbitrary custom context attributes from inject through to extract. Previously the only way to do that was to encode them in TraceContext (which has length limitations and only supports string-encoded data).

Additionally this makes the spec simpler because using a Context as a parent was already required for using Propagators.Extract, and getting a Context containing a certain Span in some way was already required for using Propagators.Inject. This PR removes the need to deal with parent SpanContexts and Spans in addition.

~~See motivating example of how this can be used to simplify propagator code that currently uses W3C TraceState at https://github.com/open-telemetry/opentelemetry-specification/pull/875#issuecomment-685659609~~ The example no longer works when removing storage of Context. A custom SpanProcessor.OnStart could store the custom Context properties as Span attributes, but these would not be propagated to child spans.

This would actually have been required by OTEP 66 (https://github.com/open-telemetry/oteps/blob/master/text/0066-separate-context-propagation.md#observability-api):

> Note that OpenTelemetry APIs calls should always be given access to the entire context object, and never just a subset of the context, such as the value in a single key.

We did intentionally not implement OTEP 66 fully though because it would have been too much to change. Still this PR does help reap more value from our partial OTEP 66 implementation.

## Possible follow-ups

* Should the parent Context also be stored on the Span (this PR originally required that)? Or: Put the parent Context inside the SpanContext instead of on the Span. With that we would be able to keep the APIs accepting SpanContext and Span. https://github.com/open-telemetry/opentelemetry-specification/pull/875#issuecomment-688097999
  * If we do either of that: Should W3C TraceState be removed from SpanContext? It could be hidden in the Context (maybe with a publicly accessible API exposed in the W3C propagator)?
* Should APIs that create and operate on Context instead of Span be added to OpenTelemetry? This could be done gradually.
* Suggested by @anuraaga in https://github.com/open-telemetry/opentelemetry-specification/pull/875#issuecomment-685418266 Add a `Map<Key<T>, T>` or similar to the `SpanContext` for arbitrary non-string extra data -- this would shield users a bit more from having to deal with Context (although they still have to do that when using propagators) while still allowing tracing propagators to transport arbitrary information. It would, however not help with connecting metrics or baggage to spans.
  + @anuraaga [notes](https://github.com/open-telemetry/opentelemetry-specification/pull/875#issuecomment-687999941) that this can be solved by copying the baggage into that map.

Oberon00 commented:
CC @tedsuo I think you said in last weeks SIG meeting that this context transport from extract to inject should work.

bogdandrutu commented on specification/trace/api.md:
```diff
@@ -219,8 +219,7 @@ the entire operation and, optionally, one or more sub-spans for its sub-operatio
   `Span`
-- A parent span in the form of a [`Span`](#span), [`SpanContext`](#spancontext),
-  or null
+- A parent span in the form of a (possibly empty) [`Context`](../context/context.md)
```
We should also allow just `Span` in case people are not using Context

Oberon00 replied:
Why should people not use context? It is possible to support this but it would introduce a complication. We would have to specify: "If a Span is passed as Context, the parent Context of the child span is the parent Context of the parent span (i.e., the grandparent Context) with the Span in it replaced with the parent Span".

bogdandrutu replied:
In JS I know it is a problem

bogdandrutu replied:
Also in some cases you start an operation and directly interact with the Span.

Oberon00 replied:
Context is already the only way to inject/propagate a Span. So both of the problems you mention are necessarily already solved by all SIGs that implement the OTEP66 propagation spec.

dyladan replied:
> In JS I know it is a problem

It's not a problem in JS. The issue in JS is that automatic propagation is a little bit wonky, but this doesn't change that.

bogdandrutu replied:
@Oberon00 believe me or not, I saw lots of cases where just `Span` could be provided (no longer work at Google to show more examples). I do believe this is a mistake to not support Span as well. Don't know how to convince besides waiting for a language to implement this and make all changes to all the instrumentations to see there will be cases where this is not possible.

Oberon00 replied:
In these cases, you can also not inject. Or rather, you can by using something like TracingContextUtil.currentContextWithSpan. This at least makes it more explicit that you are dropping information (and more cumbersome, which I think is OK here).

tsloughter replied:
I disagree with this and think it should be only a Span/SpanContext. Making the parent a Context means there is an association between the Span and the Baggage. I may be wrong but my understanding was that they are separate concerns.

If the issue is returning to the parent but not losing the baggage (or anything else in the context) then the fix should be to not reset the entire context.

In the Erlang implementation the Context contains a list of SpanContexts. When the current span is ended it is popped from the head of the list, the rest of the Context remains unchanged.

You could call the list of SpanContexts the TraceContext, which is just one part of the total Context.

Oberon00 replied:
> my understanding was that they are separate concerns.

My understanding was that while they are separate concerns, we do want to be able to cross-reference and connect different concerns, and the Context should be the place where everything comes together. That was how I understood OTEP66, and also our current propagator spec. But I guess this needs a decision. If we don't want this connection (#867, #381 seem to imply we do), then I agree that this PR is not the way we should go and @anuraaga's idea of extending SpanContext is better.

bogdandrutu replied:
I think, I will self apply the rule of "minimal api" :). We know for sure we need the "context" parent so it is great, for the "Span" parent we may or may not have good use-cases, so maybe add it later if that is the case.

So happy for me to start with only accepting Context for this because we can add later the "Span".

Oberon00 commented:
Updated the rationale in the PR description to add:

> This makes the spec simpler because using a Context as a parent was already required for using Propagators.Extract, and getting a Context containing a certain Span in some way was already required for using Propagators.inject. This PR removes the need to deal with parent SpanContexts and Spans in addition.

carlosalberto commented:
As this (probably) may end up being a non-trivial change, I'd like to see a prototype. Maybe in Java? We already have `setParent(Context)` overloads for `Span` and `CorrelationContext` creation.

Oberon00 commented:
> As this (probably) may end up being a non-trivial change, I'd like to see a prototype.

Fair point. I will be on vacation the next few days, but will try to work on it on Tuesday (of course if anyone else wants to prototype ahead, please do so and I'll happily review the prototype later).

carlosalberto commented:
> Actually I don't expect any problems in Java

Although that seems to be the case, I'm afraid of non-obvious paths, and also that some scenarios may end up being too funky. So yes, let's try that out next week, when you come back :) (unless somebody wants to take a stab at it)

Oberon00 commented:
I created a prototype for Java at https://github.com/open-telemetry/opentelemetry-java/pull/1611

Oberon00 commented:
@carlosalberto I don't fully understand the issues with baggage in the OpenTracing shim, but would putting the baggage in the context become easier with this PR? #768

anuraaga commented:
When reviewing https://github.com/open-telemetry/opentelemetry-java/pull/1611, came up with these high level comments. I think the code demonstrates that a lot of complexity is added (deleting methods is reducing API surface, but the complexity of the code went up a lot which isn't captured in the lines diff) and couldn't see us getting much from it.

I feel this is conflating two unrelated ideas, the parent / child relationship of spans (a logical connection of IDs) and a mechanism for propagating data within a process, often across threads (Context). The end result is contexts have spans, and spans also have contexts - it's not quite circular but it's not a logical relationship I think.

If someone needs to propagate data along with the span, there's no need to do it with as part of parent/child relationship I think, the data can be added to the Context next to the span, e.g., after creating the span, `Context.withValue(valuetopropagate, Context.current())`. If we really need a simpler API for adding arbitrary information to go along with the span (that is how I'm reading the rationale in #875), we could have a first-class field for extra data like [Brave](https://github.com/openzipkin/brave/blob/master/brave/src/main/java/brave/propagation/TraceContext.java#L196). But I find it confusing if we try to use the span parent for this sort of data.

Oberon00 commented:
> the code demonstrates that a lot of complexity is added

I think the complexity is mostly added by changing existing unit tests. Real code will mostly deal in (current) contexts anyway. I do not see a fundamental reason why this would increase complexity.

Oberon00 commented:
> data can be added to the Context next to the span, e.g., after creating the span, Context.withValue(valuetopropagate, Context.current())

That's exactly what this PR enables. Before this, such data would not be passed to samplers or exporters. Also, users would easily forget to propagate the context cross-thread (/cross-async task) since the API currently makes it the default to propagate only a parent span instead of the full parent context.

Oberon00 commented:
This is actually required by OTEP 66 (https://github.com/open-telemetry/oteps/blob/master/text/0066-separate-context-propagation.md#observability-api):

> Note that OpenTelemetry APIs calls should always be given access to the entire context object, and never just a subset of the context, such as the value in a single key.

By giving only the span instead of the entire context, we work against the spirit of OTEP 66.

anuraaga commented:
But `Context` itself has a parent-child like relationship, everytime the context is updated it is forked off of the current context. Failing to do this in instrumentation for example would cause far more serious bugs than losing custom values set in the context. e.g.

```java
Span parent;
Context parentContext = TracingContextUtils.withSpan(parent, Context.current());
parentContext = parentContext.withValue(customKey, customValue);

try (Scope scope = ContextUtils.withScopedContext(parentContext)) {
  Span child = Span.builder().start();
  try (Scope childScope = TracingContextUtils.currentContextWith(child)) {
    // Works, propagator will get it too here
    // https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/HttpClientTracer.java#L80
    customKey.get(childContext)  
  }
}
```

It's true this only handle propagators, not exporters. But I think sending arbitrary Context to exporters is semantically fishy anyways - the context can be used by anyone, it's not only tracing, so exporters would receive data not intended for them. Having a first class API on the `SpanContext` itself for storing data instead of piggy-backing on Context seems like a simpler, and more direct way of accomplishing propagation to exporters right?

Oberon00 commented:
> Having a first class API on the SpanContext itself for storing data instead of piggy-backing on Context seems like a simpler, and more direct way of accomplishing propagation to exporters right?

I don't know. Why add an additional value to SpanContext? I would rather prefer removing TraceState from Context instead. Then the spancontext would be back to a simple immutable value type.

Also, adding stuff seems to increase complexity more than removing/changing stuff. So I don't really believe that this would be simpler or more direct.

anuraaga commented:
> Why add an additional value to SpanContext?

Since it's explicitly tied to the span itself - context lifecycle is so far apart from span lifecycle (user can do anything with the context) that it really seems to make the data hard to reason about. If someone wants to add data to a span, they should add it to the span - contexts aren't spans. I think this alludes to @bogdandrutu's point that we do run into situations that deal with spans, our native data type, and need to accept them, and we can treat contexts as a utility.

If `TraceState`'s restrictions are causing problems, how about lifting them? `TraceState` could be a map of `TraceStateKey<T> -> T` where `TraceStateKey` defines the serialization to String. The restrictions seem like they should only apply when serializing to w3c anyway, not to the data model itself. API-wise it's very similar to `Context` but without being a thread-local, in-process propagation mechanism, just a way to store data on the Span as it travels around, which I believe is what we want here.

Oberon00 commented:
> we do run into situations that deal with spans, our native data type, and need to accept them

OTEP 66 states in https://github.com/open-telemetry/oteps/blob/master/text/0066-separate-context-propagation.md#Observability-API: 

> Note that OpenTelemetry APIs calls should *always* be given access to the entire context object, and never just a subset of the context, such as the value in a single key. This allows the SDK to make improvements and leverage additional data that may be available, without changes to all of the call sites.

(emphasized *always* in original)

Oberon00 commented:
> If TraceState's restrictions are causing problems, how about lifting them? TraceState could be a map of TraceStateKey<T> -> T where TraceStateKey defines the serialization to String. The restrictions seem like they should only apply when serializing to w3c anyway, not to the data model itself.

That's an alternative design I could get behind too, but IMHO a larger conceptual change and also not following OTEP66's spirit. I agree with OTEP66 that passing the full context is more future proof.

Conceptually I view every API object (like Span) as being just an entry point for API calls while the actual data is manipulated in the Context.

anuraaga commented:
I realized that context APIs don't allow iterating over the context, so my concern of leaking unrelated information into the exporter isn't really a concern.

So still about lifecycle, how about an example with the root span. Want to propagate some information with it.

```
Context context = Context.current().withValue(cat, meow);
Span span = tracer.spanBuilder().parent(context).start();
```

I want to add data to the span, so I have to add it to the context. And I add a parent even though it's a root span. This API seems very confusing to me. Perhaps builder has helpers to set the values, but then this would create an internal context forked from current when building the span that probably can't be mounted. I think we would have to change `start` to return `Context` instead of `Span` with such an approach to make clear that the old context needs to be invalidated by propagating this new one. This might make some sense and makes me more comfortable with the change, but it seems like a huge change too.

Oberon00 commented:
> I want to add data to the span, so I have to add it to the context. And I add a parent even though it's a root span. This API seems very confusing to me.

The typical use case for adding custom information will be the extractor. There is no span available, but you have to return a Context anyway, so it's natural to add information there. If you have a root span, you typically can add information as a span attribute -- unless you want it to be propagated, in which case you do indeed need to use the Context. But I don't think this is all that confusing.

>  I think we would have to change start to return Context instead of Span with such an approach to make clear that the old context needs to be invalidated by propagating this new one.

Indeed, OTEP 66 did specify start as:

> The following are notes on the API, and not meant as final.

> **`StartSpan(context, options) -> context`** When a span is started, a new context is returned, with the new span set as the current span.

But since I don't think the API of using Context explicitly is that confusing, I also don't see a huge need to modify that.

> This might make some sense and makes me more comfortable with the change, but it seems like a huge change too.

You can add it in a backwards-compatible way. For example, in Java you could have a function `witValue(key, value)` on the builder that returns a new builder type which has a `startContext` instead of `start` which returns a Context. But as said, I don't think this is really necessary.

Oberon00 commented:
A motivating example is probably in order. In Dynatrace we currently have something like this in Extract:

```java
private static final HttpTraceContext httpTraceContextFormat = new HttpTraceContext();

@Override
  public <C> Context extract(Context context, C carrier, Getter<C> getter) {
    // Note: 
    Context extracted = httpTraceContextFormat.extract(context, carrier, getter);
    final SpanContext value = TracingContextUtils.getSpan(extracted).getContext();
    DynatraceLinkInfotagInfo = DtLinkSerializer.parseTraceState(value.getTraceState(), tenantId);
    TraceState traceState = value.getTraceState();
    if (tagInfo == null) {
      tagInfo = new DynatraceLinkInfo().generateRootPathRandom();
      traceState =
          traceState
              .toBuilder()
              .set(
                  DtLinkSerializer.getTraceStateKey(tenantId),
                  DtLinkSerializer.toTraceStateValue(tagInfo))
              .build();
    }
    return TracingContextUtils.withSpan(
        DefaultSpan.create(
            SpanContext.createFromRemoteParent(
                value.getTraceId(),
                value.getSpanId(),
                value.getTraceFlags(),
                traceState)),
        extracted);
  }
```

We do need this rootPathRandom in the exporter, which currently again parses it from the traceState.
For different reasons, `Inject` also needs to manipulate a value on the tracestate so it also has to parse the it again, which means that for each span we have to parse a tracestate entry when we export it, and again for each time a span is injected and extracted. In the future, we might additionally need to parse the TraceState in the sampler. Logically, parsing only an incoming tracesate value on inject should be enough.

If we could access the Context in the exporter, this would be much simpler:

```java
  public static final Key<DynatraceTagInfo> DT_TAG_INFO_KEY =  Context.key("DT_TAG_INFO_KEY")

  @Override
  public <C> Context extract(Context context, C carrier, Getter<C> getter) {
    Context extracted = httpTraceContextFormat.extract(context, carrier, getter);
    final SpanContext value = TracingContextUtils.getSpan(extracted).getContext();
    DynatraceTagInfo tagInfo = DtLinkSerializer.parseTraceState(value.getTraceState(), tenantId);
     if (tagInfo == null) {
       tagInfo = new DynatraceLinkInfo().generateRootPathRandom();
     }
     return extracted.withValue(DT_TAG_INFO_KEY, tagInfo);
  }
```

If we were able to manipulate a `Map<Key<T>, T>` on the SpanContext, we would be able to avoid the performance hit but would still need to do do the dance of "get span and spancontext out of context, create a new span with a new context with the map replaced on the span context, create a new context with this span".

anuraaga commented:
@Oberon00 Definitely agree the current method of using `TraceState` is not a good one for this so we need to find a good solution for it. 

> But as said, I don't think this is really necessary.

I don't know about this - looking at the API in the Java prototype, I think it is too confusing to understand what is a span vs context if we do take a sort of half-implementation of OTEP 66. If a context is the one source of truth and we need to make sure we connect with it in tracing, the presence of the `Span` causes both confusions and I suspect potential for bugs. If there's no `Span` at all (just a hidden away implementation detail at best), and all the ways to interact with a span are methods that operate on context, e.g.

```java
class Tracer {
  void setAttribute(Context context, String key, AttributeValue value) {
    TracingContextUtils.getSpan(context).setAttribute(key, value);
  }

  void end(Context context) {
    TracingContextUtils.getSpan(context).end();
  }
}

// Span, TracingContextUtils, and similar are all private to the SDK, users only interact with Context - Context lifecycle is guaranteed to correspond with Span lifecycle.
```

Then things are very clear! It seems to be what OTEP 66 aims for, but it's a huge change in model. I'd be for the change if there was buy in, but I'm hestitant about mixing context-native vs span-native models together which I think this PR does - experienced instrumentation writers should be ok, but if users have to write some manual instrumentation I really don't see them understanding this difference between context / span.

Oberon00 commented:
I think you may have a point but currently I'm not seeing it 😃  Can you give an example of what you think would be particularly complicated with the current API? Maybe it could be made easier in Java by allowing a Scope (returned by withSpan et al) as parent?

Oberon00 commented:
> if we do take a sort of half-implementation of OTEP 66

IMHO we already did that when changing propagators to work on Context instead of SpanContext. We have increased the difficulty of injecting a span significantly, and for no real benefit IMHO due to this half-heartedness.

Oberon00 commented:
> It seems to be what OTEP 66 aims for, but it's a huge change in model. I'd be for the change if there was buy in

I'd be all for it. I think this would make our API surface much more evolvable. It might also make things easier for auto-instrumentations

Oberon00 commented:
@open-telemetry/technical-committee I'm sure that would mean delaying the release of OpenTelemetry by some weeks, so I don't think it is possible to improve the OTEP66 implementations?

Nonetheless, I maintain the opinion that this PR is beneficial without doing any further steps into the direction of OTEP 66.

carlosalberto commented:
> I'm sure that would mean delaying the release of OpenTelemetry by some weeks, so I don't think it is possible to improve the OTEP66 implementations? I put this on the SIG Spec agenda, but comments before this would be appreciated.

I think i's too late to even try this - this wouldn't even be an easy, short discussion as it will change the model so much (this was the reason why I didn't pursue that in the resulting OTEP-66 API). There are already related approaches (as not exposing `Span`s directly), and we still are not *completely* sure that this approach works fine for all cases (people still need to review/study/comment your prototype, for example).

In summary: this sounds like something we need to try out extensively and sounds like a thing for the next major version (along other aspects that look nice but seem tricky).

**However**, exposing only `Context` as parent for `Span` / `Baggage` is something we can still consider.

Oberon00 commented:
That's what I thought. I will remove the discussion-intensive part from the SIG agenda, but as said, I'd still like us to consider merging this PR as it is now. I did do a protoype over in Java, and the changes are not *too* big: https://github.com/open-telemetry/opentelemetry-java/pull/1611

anuraaga commented:
@Oberon00 I think the biggest concern is when writing tracing instrumentation, you expect to work on a `Span`. It's what you see in the UI, it's what you're trying to generate. A `Span` is connected to another `Span` - introducing another concept `Context` as something that isn't equal to a `Span`, it's only how we form parent/child relationships, is technically feasible but doesn't follow the mental model of tracing. I feel this is the same as what @carlosalberto mentions - without verifying the feasibility of going all in on `Context` from a user perspective, to require it in parts of the API. One example that comes to mind is something like this

```java
Span backendRequest 1 = spanBuilder.start();
client.execute().thenCompose(
  Span backendRequest2 = spanBuilder.start();
  return client.execute().thenAccept(backendRequest2.end();
).thenAccept(backendRequest1.end());
```

Naturally this won't connect backendRequest2 to backendRequest1, but the user probably does understand that spans have parents (they see it in the UI). If they notice `setParent(Span)`, it's very intuitive to solve the problem, `setParent(backendRequest1)`. If they only see `setParent(Context)`. What's a `Context`? Finds in javadoc "to get the current context, call Context.current()". So they try `setParent(Context.current())` and it still doesn't work. Finally, they realise this unintuitive and incorrect way of wrapping a span, `setParent(TracingContextUtils.withSpan(backendRequest1, Context.current())`  - from an API design perspective, this indirect invocation is not clean at all, nor does it solve the problems of context propagation since this context will not actually be connected to the parent, it's connected to the root. Removing APIs that accept `Span` for parent can reduce context propagation issues but can't eliminate them as long as the APIs bridging spans to context are public meaning a user has a good chance of doing it wrong.

Javadoc on `setParent` can help with this but it will probably be at least a page of documentation to explain context propagation, a [difficult topic](https://docs.google.com/document/d/1dKXHYH1CQm4IQuRl5NZaoh2vETjRlBfhcsciz_Xs5hQ/edit?disco=AAAAJTIxwfM&ts=5e8cd479&usp_dm=false) in itself. But that's still better than nothing - if restricting parent to only be a `Context` is the only way to solve the problem of attaching information to a Span in a generic way, then it is what it is and we'll just have to make sure the documentation is thorough.

I think the reality is OpenTelemetry is tracing-first in that it is the first GA product and has been the base for our current work. That's why things started span-native - span-native isn't so bad since it's still not hard to model cross-cutting concerns with it, for example including a bag of objects in `SpanContext`, an approach taken by Brave and opentracing FWIU. Switching to context-native, as in OTEP-66, is interesting as it does generalize the concept of context outside of tracing, but as we don't have time to move completely from our span-native history, we've found ourselves in a middle-ground that seems worse than both the span-native legacy or possible context-native future.

But again, it is what it is. I think given no changes to any of our other APIs, restricting parent to a `Context` is the only way to solve the problem we have here. I think another possible approach is to add a bag to the `SpanContext`, whether as a generalization of `TraceState` to not just be a logic-version of the HTTP header but a generic bag, or as a separate generic bag structure - it brings back some of our [span-native](https://github.com/openzipkin/brave/blob/master/brave/src/main/java/brave/propagation/TraceContext.java#L196) [roots](https://github.com/opentracing/opentracing-java/blob/master/opentracing-api/src/main/java/io/opentracing/SpanContext.java#L59). I personally find the latter to be a bit more intuitive within our current, mostly span-native, API, probably for the same reasons that prevent us from going forward too quickly on OTEP-66, requiring a deeper understanding of Context by users. And I guess the latter has less chance of losing Span information if the user recreated a `Context` instead of propagating it correctly, so maybe more robust?

> IMHO we already did that when changing propagators to work on Context instead of SpanContext. We have increased the difficulty of injecting a span significantly, and for no real benefit IMHO due to this half-heartedness.

I really agree with this, I find the propagators to already be much harder to understand and we probably are on the same page there.

Oberon00 commented:
@anuraaga 

> I feel this is the same as what @carlosalberto mentions

@carlosalberto himself suggested that change in #510. We were talking about going full OTEP 66 here (e.g. returning a Context from startSpan), see my comment https://github.com/open-telemetry/opentelemetry-specification/pull/875#issuecomment-686389885 above his https://github.com/open-telemetry/opentelemetry-specification/pull/875#issuecomment-686507172.

>  I think given no changes to any of our other APIs, restricting parent to a Context is the only way to solve the problem we have here. I think another possible approach is to add a bag to the SpanContext

Since you need a Context to propagate a Span cross-process using Propagators already, this adds that requirement symmetrically for cross-thread (cross-execution context). The default case (inside a single thread/execution-context) should be to not use setParent anyway but using withSpan with the parent. So I think the Context way is more consistent.

So maybe it makes cross-thread propagation a bit harder but the total complexity of "dealing with cases where I need an explicit parent + dealing with Propagators" stays the same for users or is even reduced IMHO.

In Dynatrace we even often view instrumentations for propagation (that injects the current Context into an outgoing request, or propagates it to another thread or async task) as separate things from "normal" instrumentations that create Spans and attach attributes to them. These two things might be written by different authors in some cases even.

> Removing APIs that accept Span for parent can reduce context propagation issues but can't eliminate them

So you agree this change helps reducing issues? 😉

@anuraaga From your comment I'm not sure what your opinion on this PR is now? Do you have any particular change suggestions? Do you approve of the PR (saying that while you don't really like it moves the context/propagation/parent stuff into a "less bad" direction)? Do you think this is something we should not do at all?

Oberon00 commented:
One other idea we should have tried (but it's to late now) is SpanContext == Context (we had https://github.com/open-telemetry/oteps/issues/73 "Span == SpanContext" and https://github.com/open-telemetry/oteps/issues/68 "Remove Span" already)

anuraaga commented:
> Removing APIs that accept Span for parent can reduce context propagation issues but can't eliminate them

> So you agree this change helps reducing issues? 😉

Heh yeah that wasn't a clear statement. I think I am thinking this is a reason to not rely on `Context` in the first place - propagating a span seems to be more reliable than context in our current API (can't recreate a span out of thin air like one can with Context).

> I think another possible approach is to add a bag to the SpanContext, whether as a generalization of TraceState to not just be a logic-version of the HTTP header but a generic bag, or as a separate generic bag structure - it brings back some of our span-native roots. 

I think my suggestion is still to try this approach to have less reliance on `Context`.

> In Dynatrace we even often view instrumentations for propagation (that injects the current Context into an outgoing request, or propagates it to another thread or async task) as separate things from "normal" instrumentations that create Spans and attach attributes to them. These two things might be written by different authors in some cases even.

I think there is definitely clear separation for remote propagation. But I think there will be strong overlap between the "normal" instrumentation that create Spans and doing in-process propagation of the span for non-remote, there isn't even a `Propagator` involved here.

Oberon00 commented:
Interestingly, when looking at the Java auto instrumentation the most common explicit parent for a span seems to be `TRACER.currentSpan()` (e.g. https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/1217882a306f86/instrumentation/servlet/servlet-common/src/main/java/io/opentelemetry/instrumentation/auto/servlet/dispatcher/RequestDispatcherInstrumentation.java#L96) which looks like an anti-pattern (let's just use the default parent).

Some places even start with a context and convert it to a span using `TracingContextUtils.getSpan(context)`, e.g. https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/1217882a306f86/instrumentation/aws-sdk/aws-sdk-2.2/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v2_2/AwsSdkClientTracer.java#L94

Of course everything using the "typedspan" module in the Java auto-instrumenation does not even have an API to use Context as parent (yet), which would have to be adjusted.

While I haven't done a full quantitative analysis, my feeling is that also instrumentations would get a net benefit and reduction in complexity when applying this PR.

I have heard numerous times that there "might" be cases where you don't have a Context but only a Span. I do not believe this can logically be the case. Some refactoring might be required to store a Context instead of a span in some places (e.g. [InstrumentationContext](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/1217882a306f86/auto-api/src/main/java/io/opentelemetry/instrumentation/auto/api/InstrumentationContext.java) usages)

anuraaga commented:
My point isn't technical, it's conceptual and concerned with creating a fit API. Currently it's too easy to miss out on what context is.

> but it's not like we interoperate with some legacy systems that only give us spans.

We do - OpenTelemetry Tracer gives us spans, not contexts :) Because we interoperate with this legacy system, we have to use unintuitive concepts like wrapping it into a context. You've pointed out many points in our instrumentation that is messy because of the divide - I don't want less experienced users to have to reason this out if possible.

On the flip side, I haven't heard anything about pros / cons about my suggestion of sticking with adding a bag to the span itself. Is it such a bad idea? 😅

Oberon00 commented:
I think your suggestion with the map is fine if we only consider tracing. However, I think we have a better shot at resolving issues such as #867 and #381 that aim to associate other aspects with tracing later, if we just bite into the context bullet.

Oberon00 commented:
> We do - OpenTelemetry Tracer gives us spans, not contexts

Wouldn't you normally use something like `Tracer.withSpan` right after creating that though? Maybe we just need to also add the right utility APIs, like a `Context SpanBuilder.startAndCaptureCurrent()` and a `ScopedSpan TracingContextUtils.withCurrentContext(Context)`.

anuraaga commented:
> Wouldn't you normally use something like Tracer.withSpan right after creating that though?

I hope so since I have worked with our strange API. But I might forget and our instrumentation has often forgotten. And to this day I can _never_ remember if the API I'm looking for is on TracingContextUtils or ContextUtils. With context-native, there would be no TracingContextUtils - but we're already stuck with this strange API which an expert (if I'm allowed to call myself one) gets wrong daily.

I don't think we can assume everyone using these APIs is an OpenTelemetry expert - that will cease to be the case within months after GA. So I'm looking at this in terms of creating an API users enjoy using - providing a lot of helpers with even more new terms like scopes isn't so fun. If tracer returns a span, setParent must accept a span. I think this philosophy makes sense in API design, symmetry is important, even if it may result in some corner cases.

I've wrung myself dry in what I can add to this discussion I think. In the end I'm not a spec approver and hope to hear more thoughts from others, hopefully after taking a step back.

Oberon00 commented:
I think we can add a convenience API to accept a Span additionally. I don't like it but it may make it easier to do the only slightly wrong thing easier (and sometimes even the right thing) if we had a method `SpanBuilder.captureCurrentAndSetParent(Span)` that captures the current context immediately and puts that span into it.

Oberon00 commented:
> I've wrung myself dry in what I can add to this discussion I think. In the end I'm not a spec approver and hope to hear more thoughts from others, hopefully after taking a step back.

Same here. I think we have brought up many important points in our discussion so far but it's now time for others (@open-telemetry/technical-committee, @open-telemetry/specs-approvers , @open-telemetry/specs-trace-approvers  and also users like @open-telemetry/java-instrumentation-maintainers) to comment.

tsloughter commented:
Initial reaction to the title is, no, they should continue to have Parent SpanContexts and instead Context must be what is operated on/returned by the Tracer (or whatever module/object puts spans in and out of active context).

But I still need to read all the past discussion in this to give a more complete response.

Example in erlang-ish psuedo code:

```
// set a correlation on the current context which is a "thread local"
// this function probably returns a context token or the full context but here we won't use it, so it is ignored
_ = ot_correlations:set(key-1, value-1)

// Ctx has SpanCtx as the active Span and Ctx is also the active context in the "thread local" 
{SpanCtx, Ctx} = ot_tracer:start_span(span-1)

// a new context (or token to the context) is returned while the active context in the "thread local" is unchanged
// note that SpanCtx1 *is* the active span in Ctx1
{SpanCtx1, Ctx1} = ot_tracer:start_inactive_span(span-2)

// create a new process/thread/coroutine and set the context
spawn(fun() ->
        opentelemetry:set_context(Ctx1)

        // prints value-1
        Value = ot_correlations:get(key-1)
        print(Value)
      end) 
````

Here `Ctx1` has a copy of the correlation context from `Ctx` and has `SpanCtx1` as the active span, so `set_context` results in `SpanCtx1` being the active span in the "thread local" of the new process/thread/coroutine.

So the key difference in this is that `Ctx1` which is returned by `start_inactive_span` includes the new span as its active span, this would be something new as far as I know.

tsloughter commented:
I realize now that getting/setting the entire context means losing correlations in a common case where you want to wrap a call to user code from your library -- like an OpenTelemetry webserver instrumentation that starts a span and then calls the users request handling code -- with an exception handler to end the outer span and reset the context to what it was before the span. 

But you probably don't want to lose the correlations that may have been updated in the user code?

```
Ctx0 = opentelemetry:get_context()

// set a correlation on the current context which is a "thread local"
// this function probably returns a context token or the full context but here we won't use it, so it is ignored
_ = ot_correlations:set(key-1, value-1)

// Ctx has SpanCtx as the active Span and Ctx is also the active context in the "thread local" 
{SpanCtx, Ctx} = ot_tracer:start_span(span-1)

try
  ot_correlations:set(key-2, value-2)
after
  // end the first span
  ot_span:end_span(SpanCtx)
  // reset the context
  opentelemetry:set_context(Ctx0)
end

// should this key be defined and return value-2?
ot_correlations:get(key-2)
```

I had assumed this was a unique issue to Erlang/Elixir because we rely on a list of spans in the context so that ending can return to the previous span, while other languages have some scoping functionality where they can always perform an action on the current span when leaving a scope, but this issue and another discussion makes it seem there is a similar issue across languages around keeping span and correlation contexts in sync and what "in sync" means.

tsloughter commented on specification/trace/sdk.md:
```diff
@@ -202,12 +202,19 @@ Thus, the SDK specification defines sets of possible requirements for
+  (implicitly) associated with the span.
+  
+  Note: The `Context` of the span itself cannot be made available,
+  because a Span may be (active) in several or no `Context`'s at any time.
```
I did not think a span could be active in several contexts?

Oberon00 replied:
I somehow missed this comment. But yes, a span can be active in as many contexts as you want. Why should it not? E.g. consider a operation that is parallelized over multiple threads (or even if one thread waits for another): The span could be active in both at the same time. There is no link back from Span to Context, so this should not be a problem.

anuraaga commented:
@Oberon00 Thanks for the summary! 

> It would, however not help with connecting metrics or baggage to spans.

Just want to clarify, with a `Map` on the `SpanContext`, I was thinking baggage would also be copied into it.

Oberon00 commented:
> I was thinking baggage would also be copied into it.

Hmm, then the SpanContext would basically become a TelemetryContext. I wonder if we would instead just have a full Context inside the SpanContext, we would need to change nothing in the tracing APIs. We would just have to establish rules how to connect the SpanContext to a Context, add a getContext and be done with it. No user-facing API changes needed. If we can find good rules for that connection, that might just be the perfect solution. 🤔

Oberon00 commented:
Actually the rules for associating a Context with a SpanContext would be dead simple: We have an explicit parent Context, SpanContext or Span already everywhere except when starting a new Span, where we already have rules for determining a parent SpanContext which again contains a Context. So we can basically use that Context for associating with the SpanContext.
There is only one question. The SpanContext will never itself be in its associated Context (it can't because both Context and SpanContext are immutable). When associating a Context with a SpanContext by reading it from a parent SpanContext, we could create a new context that is modified by putting that parent SpanContext into the Context. I'm not sure if that would be useful though.

tsloughter commented:
> associating a Context with a SpanContext

A SpanContext should be associated with a Context, not the other way around. This is creating a circular dependency and confusing stacking where a Context contains a SpanContext which contains a Context.

Oberon00 commented:
I agree that this is a bit confusing. That's why I'd prefer the current PR.

@anuraaga [suggests](https://github.com/open-telemetry/opentelemetry-specification/pull/875#issuecomment-687999941) adding a `Map<Key<T>, T>` to the SpanContext and

> Just want to clarify, with a Map on the SpanContext, I was thinking baggage would also be copied into it.

what do you think of that suggestion?

tsloughter commented:
Possibly. It is a fairly large change in a sense (not in the code) so would want to think it over more.

Would the Map be just the Baggage? Or is it general, like a context map, that which would then have a Baggage key pointing to the map of its key/values.

Oberon00 commented:
IMHO we definitely need a way to transport arbitrary values across explicit parents. So it would be a general map.

carlosalberto commented:
>  I wonder if we would instead just have a full Context inside the SpanContext, we would need to change nothing in the tracing APIs.

That's not possible, as a `Span` may be active in different threads/execution-units and thus be active in different `Context` objects.

carlosalberto commented:
> @anuraaga suggests adding a Map<Key<T>, T> to the SpanContext 

Different discussion needs to happen. Maybe fill an issue to do a follow up for that? (For the record, I was interested in adding something like this to `Span`, although for different technical reasons ;) )

Oberon00 commented:
@carlosalberto 

> > I wonder if we would instead just have a full Context inside the SpanContext, we would need to change nothing in the tracing APIs.
> 
> That's not possible, as a `Span` may be active in different threads/execution-units and thus be active in different `Context` objects.

Sure it is possible, but you would have a "parent" Context on the SpanContext. The current context cannot be on the SpanContext of course.



> > @anuraaga suggests adding a Map<Key, T> to the SpanContext
> 
> Different discussion needs to happen. Maybe fill an issue to do a follow up for that? (For the record, I was interested in adding something like this to `Span`, although for different technical reasons ;) )

I think this is the same discussion. If we accept this PR, we don't need that map IMHO, as Context can be used instead.

tsloughter commented:
I'm actually not getting parts of the issues this is supposed to fix. What is wrong with working with Contexts. inject and Extract take from or insert to the current Context. 

> using a Context as a parent was already required for using Propagators.Extract, and getting a Context containing a certain Span in some way was already required for using Propagators.Inject.

Why are you "getting a Context containing a certain Span", it should be whatever Span is active in the current Context when you are injecting.

Why is a parent Context needed for Extract? Doesn't it simply update the current Context but puts the SpanContext that is Extracted into a special Context key for "remote span context"?

tsloughter commented:
From the spec meeting I wanted to re-iterate that I think the issue is use of Tracer's `set_span` instead of context `attach`.

Starting a span with a parent that is a Context implies that the Tracer must know to set baggage, right? And that is a link that would mean anything new in Context would have to update Tracer code to be able to do the same. I think it also risks losing baggage in metrics code because the user is used to moving baggage around only with the Tracer and Span.

So the code should be like:

If you have a Context (`Ctx`) you would pass as the Parent in the case of Span starting allowing a Context to be passed you would instead do:

```
context.attach(Ctx)
tracer.start_span("some-span", #{parent: tracer.get_span(Ctx)})
```

If the point is to send this child span off to another thread and not permannently change the baggage in the current context:

```
// make the context we want to use as the parent of the new span the current context
OriginalCtx = context.detach(Ctx)

// create an active span on what is now the current context
InactiveChildSpan = tracer.start_span("some-span")

// set the current threads context back to the original and get a copy of the context we want to pass to the new thread
ChildCtx = context.detach(OriginalCtx)

// create new thread/process/async callback/whatever and attach this context
// this makes the current span of the new thread "some-span"
spawn(fun() -> context.attach(ChildCtx) end)
```

No need to pass Context as part of `start_span`.

Oberon00 commented:
> Starting a span with a parent that is a Context implies that the Tracer must know to set baggage, right?

I don't see how? The baggage would jus "exist" on the Context, the tracer doesn't need to know anything about it. The exporter does it if it wants to export it.

> I think it also risks losing baggage in metrics code because the user is used to moving baggage around only with the Tracer and Span.

You mean they won't set the right context as active when recording metrics?

Oberon00 commented:
One thing that was suggested (even back in OTEP 66) was to have StartSpan return a Context. I could then imagine that we extend this to only allow getting an Span API object while the span's Context is active.

tsloughter commented:
> I don't see how? The baggage would jus "exist" on the Context, the tracer doesn't need to know anything about it. 

A Span is started through the Tracer. If you pass a Context as the parent you want to have this Context and its baggage associated with the new Span, right? How else would that be possible unless the Tracer knew this? Or the Tracer attached the Context.

> The exporter does it if it wants to export it.

Do you mean this is simply so that when exporting a Span there is a Context associated with it that Baggage can be fetched from?

What does an exporter do with Baggage? Has it been added as something that gets exported in the trace export protocol?

I thought any Baggage that was to be exported with a Span had to be specifically set as attributes on the Span with values pulled from Baggage.

Oberon00 commented:
> Or the Tracer attached the Context.

This is what the PR proposes 😃  The readable Span gains a getParentContext method that returns the full parent context.

Oberon00 commented:
> What does an exporter do with Baggage? Has it been added as something that gets exported in the trace export protocol?

No, but maybe it should. Or we might define that it gets added as span attributes (with a certain prefix?).

> I thought any Baggage that was to be exported with a Span had to be specifically set as attributes on the Span with values pulled from Baggage.

I thought that Baggage was the thing that was implicitly associated with all telemetry. And if we don't export it, what is the point of associating it?

tsloughter commented:
I don't think Baggage should be exported with every span (and metric?).

If a decision is made that it should be then that changes everything and I see the need for either associating baggage with a span so the exporting of a span can know the baggage of the span at the time of its creation instead of the current Context baggage.

But I really don't think we should go that route and so disagree there is a reason to couple a span and baggage.

Oberon00 commented:
Discussed in the SIG meeting:
* Keep the API changes as-is (only allow Context as parent, no Span, no SpanContext) 
* Remove storage of Context from this PR.
* Pass parent Context to SpanProcessor OnStart instead.

Oberon00 commented:
I noticed that my main example that motivated this PR does no longer work after removing storage of Context on the Span. But we can discuss this separately (even after-ga, as it's SDK-only):  https://github.com/open-telemetry/opentelemetry-specification/pull/875#issuecomment-685659609 The example no longer works when removing storage of Context. A custom SpanProcessor.OnStart could store the custom Context properties as Span attributes, but these would not be propagated to child spans where they would also be needed in this example.

Oberon00 commented:
Java prototype was updated.

bogdandrutu commented:
@Oberon00 the context will still be propagated on the user side so things that were available in the parent span will also be available in the child Span. So not sure I understand your concern about not storing the context

Oberon00 commented:
Indeed, we can now say to the user that if the context didn't make it from extract to inject, they should fix their code. But the context is no longer available to the exporter now (unlike the w3c tracestate)

tsloughter commented on specification/trace/api.md:
```diff
@@ -391,6 +371,21 @@ The Span creation API MUST provide:
+
+The API MUST provide functionality to get the effective Span from a `Context`.
+
+Given a `Span` with its parent `Context`, the semantic parent Span is the
```
It is unclear how this was meant to be read. "Given a Span with its parent Context" sounds like it could mean that a Span has a parent Context, or is this supposed to be saying "Given a Span and its parent Context" since they are separate.

Oberon00 replied:
Indeed I forgot to remove this paragraph. It is redundant with the

> The semantic parent of the Span MUST be determined according to the rules
> described in the [Effective Span section](#effective-span).

description above in the span creation API.

Oberon00 replied:
I removed the now unneeded changes here in f98a516 and ad4a72f.

carlosalberto commented:
Waiting for more prototypes to validate this, but otherwise LGTM.

Oberon00 commented:
@bogdandrutu 

> But the context is no longer available to the exporter now (unlike the w3c tracestate)

Actually, thinking about it again, this should still be doable with a little extra work *if* we have a readable span in the parent `Context`. E.g. in Java this would work with a (checked) cast from `Span` to `ReadableSpan`. From there you could copy certain attributes to child spans, and thus use span attributes to store custom values instead of the tracestate. The initial extract would set them directly on the Context. So in `OnStart(Span started, Context parent)` you would do something like

```java
long myvalue1 = 0;
String myvalue2 = null;
Span parentSpan = TracingContextUtils.getSpan(parent);
if (parentSpan instanceof ReadableSpan) {
  SpanData sd = parentSpan.toSpanData();
  myvalue1 = sd.getAttributes().getLongAttribute("_myvendorcustom.myvalue1");
  myvalue2 = sd.getAttributes().getStringAttribute("_myvendorcustom.myvalue2);
} else {
  final MyCustomValues myCustomValues = MY_CUSTOM_VALUE_KEY.get(parent);
  if (myCustomValues != null) {
    myvalue1 = myCustomValues.myvalue1;
    myvalue2 = myCustomValues.myvalue2;
  }
}
```

which is rather ugly now that I write it out, but we might improve on that in the future.

tsloughter commented:
"no longer available to the exporter"

I think that is related to one of my questions for the meeting today. If Context is available in `OnStart` shouldn't it be available in `OnEnd`. I've argued against ending a Span with a method on the Span like `Span.End()` before and I think this shows why that is important, unless we can get away without needing Context in the processor.

Oberon00 commented:
Which context would you want to pass in end? The context in which it was ended? Then you can just use `Context.current()`. The parent Context? Then you would need to store it inside the Span, which you (and others) said we should not do.

carlosalberto commented:
I suggest we don't talk about potential improvements and additions in order to not lose focus, *unless* it's something truly, absolutely needed by the issue/PR at hand.

It looks like @tsloughter @codeboten are working on prototypes of this, so please comment/approve whenever you are able to provide your feedback :)

Oberon00 commented:
@carlosalberto If you want to wait for prototypes, I suggest you should mark this as "request changes".
